Sort products into groups in the final offer PDF printout

Steffengreiner opened the pull request:
- [X] This comment contains a description of changes (with reason)
- [X] Referenced issue is linked
- [X] `CHANGELOG.rst` is updated

**Description of changes**
This PR addresses #359 by sorting the products according to their associated services("Data generation", "Data analysis " and "Project Management and Data storage") and productName in the Offer PDF. 
Additionally it removes the calculation methods for the subtotal tables and the methods used for the previous sorting of the products. 

**Additional context**
Adding the total net sum and overhead cost as required by #360 of the newly sorted products will be addressed in a seperate PR.

KochTobi commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -180,15 +169,32 @@ class OfferToPDFConverter implements OfferExporter {
         //Generate ProductTable for Overhead and Non-Overhead Product Items
-        generateProductTable(OVERHEAD, listOverheadItems, maxTableItems)
-        generateProductTable(NO_OVERHEAD, listNoOverheadItems, maxTableItems)
+        generateProductTable("Data Generation", listDataGenerationItems, maxTableItems)
+        generateProductTable("Data Analysis", listDataAnalysisItems, maxTableItems)
+        generateProductTable("Project Management and Data Storage", listNoOverheadItems, maxTableItems)
```
```suggestion
        generateProductTable("Data Generation", listDataGenerationItems, maxTableItems)
        generateProductTable("Data Analysis", listDataAnalysisItems, maxTableItems)
        generateProductTable("Project Management and Data Storage", projectManagementItems + dataStorageItems, maxTableItems)
```
or similar. Since you do not differentiate between with overheads and without overheads anymore?

Steffengreiner commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -41,18 +42,6 @@ class OfferToPDFConverter implements OfferExporter {
      */
     static final CHROMIUM_EXECUTABLE = "CHROMIUM_EXECUTABLE"
 
-    /**
-     * Is the preceding part of the element id used in the html document for sections associated with overhead cost
-     * If the String value is changed here it also has to be adapted in the html document
-     */
-    static final String OVERHEAD = "overhead"
-
-    /**
-     * Is the preceding part of the element id used in the html document for sections associated without an overhead cost
-     * If the String value is changed here, it also has to be adapted in the html document
-     */
-    static final String NO_OVERHEAD = "no-overhead"
-
```
Good Point, however this was not part of a previous release and these variables were introduced and used only for the PDFConverter class itself

Steffengreiner commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -180,15 +169,32 @@ class OfferToPDFConverter implements OfferExporter {
+            if (it.product instanceof Sequencing) {
+                listDataGenerationItems.add(it)
+            }
+            else {
+                listDataAnalysisItems.add(it)
+            }
```
Currently the offer dto and offer entity provide properties that differentiate between overhead and no overhead items and the associated overhead and net cost for each. 
In the list of items with overhead there are 3 dto classes:

1. Sequencing
2. Primary Analysis
3. Secondary analysis

In contrast the list of no overhead items contains the classes 

1.Project Management
2. Data storage

That's why it's enough to sort out the sequencing classes from this list since it only contains item with overheads.

Steffengreiner replied:
the differentiation is a remnant from the previous iteration where it was required to provide subtables for the item grouping into overhead and no overhead association. This was done in the Offer entity and the offer dto by grouping the items into the seperate lists.
Here i just reused that distinction to get to the relevant classes. 
(I think back then the idea was to avoid having business logic in the Converter itself and keep the logic to the Offer entity.)
If you think accessing the classes in the converter makes more sense, then i'll include it in the PR 👍

Steffengreiner commented:
@KochTobi I addressed your review and removed the distinction of overhead and no overhead items. 
I also sort the items into the new Categories Data generation, data analysis and data storage & Project management in the Converter explicitly and named the variables accordingly! 
Thanks for the detailed feedback 👍

KochTobi commented:
> @KochTobi I addressed your review and removed the distinction of overhead and no overhead items.
> I also sort the items into the new Categories Data generation, data analysis and data storage & Project management in the Converter explicitly and named the variables accordingly!
> Thanks for the detailed feedback 👍

Your changes look good at the first glance. As I stated previously I do not have the focus on this atm so I will dismiss my change requests.

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -177,18 +170,41 @@ class OfferToPDFConverter implements OfferExporter {
 
-        List<ProductItem> listOverheadItems = offer.itemsWithOverhead
-        List<ProductItem> listNoOverheadItems = offer.itemsWithoutOverhead
+        List<ProductItem> productItems = offer.items
+
+        List<ProductItem> dataGenerationItems = []
+        List<ProductItem> dataAnalysisItems = []
+        //Project Management and Data Storage are grouped in the same category in the final Offer PDF
+        List<ProductItem> dataManagementItems = []
+
+        // Sort ProductItems into "DataGeneration", "Data Analysis" and "Project & Data Management"
+        productItems.each {
+            if (it.product instanceof Sequencing) {
+                dataGenerationItems.add(it)
+            }
+            if (it.product instanceof PrimaryAnalysis || it.product instanceof SecondaryAnalysis) {
+                dataAnalysisItems.add(it)
+            }
+            if (it.product instanceof DataStorage || it.product instanceof ProjectManagement){
+                dataManagementItems.add(it)
+            }
+        }
```
I think this can be more expressive. For example call a method that expects for items and return a map with items sorted by category. 

```
Map itemsByGroup = groupItems(productItems)
```

Steffengreiner replied:
Good suggestion!

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -177,18 +170,41 @@ class OfferToPDFConverter implements OfferExporter {
+            }
+        }
+
+        //Sort Lists by ProductName
+        dataGenerationItems = dataGenerationItems.sort{it.product.productName}
+        dataAnalysisItems = dataAnalysisItems.sort{it.product.productName}
+        dataManagementItems = dataManagementItems.sort{it.product.productName}
```
Has this been a requirement?

Steffengreiner replied:
It was not discussed in the stakeholder meeting but does make sense to list similiar items in the final offer. Otherwise the order is dependent on the order in which the products were added to the Offer

sven1103 replied:
and that might exactly how it is desired. I would like to leave it out for now, until we get feedback for it.

Steffengreiner replied:
Alright Done! 👍

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -177,18 +170,41 @@ class OfferToPDFConverter implements OfferExporter {
-        //Generate ProductTable for Overhead and Non-Overhead Product Items
-        generateProductTable(OVERHEAD, listOverheadItems, maxTableItems)
-        generateProductTable(NO_OVERHEAD, listNoOverheadItems, maxTableItems)
+        //Generate ProductTable for "DataGeneration", "Data Analysis" and "Project & Data Management"  Product Items
+        generateProductTable("Data Generation", dataGenerationItems, maxTableItems)
+        generateProductTable("Data Analysis", dataAnalysisItems, maxTableItems)
+        generateProductTable("Project Management and Data Storage", dataManagementItems, maxTableItems)
```
can then be: `generateProductTables(itemsByGroup)`

Steffengreiner replied:
Done!

sven1103 reviewed (changes requested):
Some minor requests

Steffengreiner commented:
@sven1103 I introduced a grouping method which maps the products into the requested categories. 
If this PR is approved i'll add the SubtotalFooter with the relevant pricing in a seperate PR.  
Thanks for your feedback 👍

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -226,60 +221,64 @@ class OfferToPDFConverter implements OfferExporter {
-                    elementId = "product-items" + "-" + tableCount
-                    htmlContent.getElementById("item-table-grid").append(ItemPrintout.tableHeader(elementId))
-                    tableItemsCount = 1
+        productItemsMap.each {key, value ->
```
Can be more expressive:
```suggestion
        productItemsMap.each {productGroup, items ->
```

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -226,60 +221,64 @@ class OfferToPDFConverter implements OfferExporter {
+                def elementId = "product-items" + "-" + tableCount
+                //Append Table Title
+                htmlContent.getElementById(elementId).append(ItemPrintout.tableTitle(key.toString()))
+                value.each {productItem ->
```
+1 for being expressive here

sven1103 replied:
-1 for using the variable `value` :)

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -226,60 +221,64 @@ class OfferToPDFConverter implements OfferExporter {
+                        tableItemsCount = 1
+                    }
+                    //add product to current table
+                    htmlContent.getElementById(elementId).append(ItemPrintout.itemInHTML(itemPos++, productItem as ProductItem))
```
Are you aware of this interface method of Collections? 

http://docs.groovy-lang.org/latest/html/groovy-jdk/java/lang/Object.html#eachWithIndex(groovy.lang.Closure)

You can simplify this code block significantly and make it more readable.

And: upgrading variable inline, like ``itemPos++`` is hard to debug and not very expressive,

Steffengreiner replied:
Thanks for the suggestion, i will use it in the future and added it 👍

sven1103 reviewed (changes requested):
Nearly there!

Steffengreiner commented:
@sven1103 I implemented your feedback 🚀

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/OfferToPDFConverter.groovy:
```diff
@@ -226,63 +221,59 @@ class OfferToPDFConverter implements OfferExporter {
+                def elementId = "product-items" + "-" + tableCount
+                //Append Table Title
+                htmlContent.getElementById(elementId).append(ItemPrintout.tableTitle(productGroup.toString()))
+                items.eachWithIndex { item, itemPos ->
+                    //start (next) table and add Product to it
+                    if (tableItemsCount >= maxTableItems) {
+                        ++tableCount
+                        elementId = "product-items" + "-" + tableCount
+                        htmlContent.getElementById("item-table-grid").append(ItemPrintout.tableHeader(elementId))
+                        tableItemsCount = 1
+                    }
+                    //add product to current table
+                    htmlContent.getElementById(elementId).append(ItemPrintout.itemInHTML(itemPos, item as ProductItem))
```
way better ! 👏

sven1103 replied:
NIT: excluding the block into an own method.

sven1103 reviewed (approved):
Nice refactoring!
